### PR TITLE
feat: [CDS-78461]: Platform Service Endpoint to get board list for canny post creation

### PR DESCRIPTION
**Pull request by ashinsabu3**
## Describe your changes

This addition introduces an endpoint under platform-service GET /canny/boards(the first of the two endpoints required for the epic SPG-6971. 

Note: Regarding the one hardcoded exception in boards-list(continue statement on the for loop if board is admin only board) , this is a gap in the Canny Application which has been escalated with their team and they have added it as a future planned feature.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52596)

### CODE REVIEW

**Comment by ashinsabu3** [comment]
trigger uts

**Comment by ashinsabu3** [comment]
trigger uts

**Comment by ashinsabu3** [comment]
trigger uts

File: dockerization/platform-service/scripts/replace_configs.sh [hunk 1]
```diff
@@ -268,6 +268,14 @@ if [[ "" != "$COVEO_TOKEN" ]]; then
   export COVEO_TOKEN; yq -i '.zendeskApiConfig.coveoToken=env(COVEO_TOKEN)' $CONFIG_FILE
 fi
 
+if [[ "" != "$prod_CANNY_BASE_URL" ]]; then
+  export prod_CANNY_BASE_URL; yq -i '.cannyApiConfig.baseUrl=env(prod_CANNY_BASE_URL)' $CONFIG_FILE
```
Comment (sourabh-awashti) [inline dockerization/platform-service/scripts/replace_configs.sh:272]: we can remove prod prefix from here, as we will directly use the value as CANNY_BASE_URL from manifest.

File: dockerization/platform-service/scripts/replace_configs.sh [hunk 2]
```diff
@@ -268,6 +268,14 @@ if [[ "" != "$COVEO_TOKEN" ]]; then
...
+fi
+
+if [[ "" != "$prod_CANNY_TOKEN" ]]; then
```
Comment (sourabh-awashti) [inline dockerization/platform-service/scripts/replace_configs.sh:275]: same comment as above

File: platform-service/build/container/scripts/replace_configs.sh [hunk 3]
```diff
@@ -266,6 +266,14 @@ if [[ "" != "$COVEO_TOKEN" ]]; then
   export COVEO_TOKEN; yq -i '.zendeskApiConfig.coveoToken=env(COVEO_TOKEN)' $CONFIG_FILE
 fi
 
+if [[ "" != "$prod_CANNY_BASE_URL" ]]; then
+  export prod_CANNY_BASE_URL; yq -i '.cannyApiConfig.baseUrl=env(prod_CANNY_BASE_URL)' $CONFIG_FILE
```
Comment (sourabh-awashti) [inline platform-service/build/container/scripts/replace_configs.sh:270]: same prefix comment

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyBoardsResponseDTO.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import lombok.experimental.FieldDefaults;
+import org.hibernate.validator.constraints.NotEmpty;
+import org.jetbrains.annotations.NotNull;
+
+@Data
+@Builder
+@JsonIgnoreProperties(ignoreUnknown = true)
+@Schema(name = "CannyBoardsResponse", description = "Contains list of Boards and their ID's from Canny")
+public class CannyBoardsResponseDTO {
+  @NotNull Integer code;
```
Comment (sourabh-awashti) [inline platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyBoardsResponseDTO.java:31]: we can remove this field, as UI does not require api code response from BE

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  private final OkHttpClient okHttpClient = new OkHttpClient.Builder().retryOnConnectionFailure(true).build();
+
+  @Inject
+  public CannyHelper(@Named("cannyApiConfiguration") CannyConfig cannyConfig) {
+    this.cannyConfig = cannyConfig;
+  }
+
+  public CannyBoardsResponseDTO getBoards() {
+    try {
+      String jsonRequestBody = "{\"apiKey\":\"" + cannyConfig.token + "\"}";
```
Comment (sourabh-awashti) [inline platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java:46]: can we extract out the logic for making api request to a separate class, such as canny client. There we can have both function one for fetching boards and one for creating post

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyBoardsResponseDTO.java (hunk 4 above)
Comment (ashinsabu3) [reply platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyBoardsResponseDTO.java:31]: Addressed this.

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java (hunk 5 above)
Comment (ashinsabu3) [reply platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java:46]: Do you mean abstract out the whole getBoards function into CannyClient or just the api request part

Comment (ashinsabu3) [reply platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java]: The canny api calls are all moved to cannyclient class

**Comment by ashinsabu3** [comment]
trigger uts

**Comment by ashinsabu3** [comment]
trigger uts

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyBoardsResponseDTO.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import lombok.experimental.FieldDefaults;
+import org.jetbrains.annotations.NotNull;
+
+@Data
+@Builder
+@JsonIgnoreProperties(ignoreUnknown = true)
+@Schema(name = "CannyBoardsResponse", description = "Contains list of Boards and their ID's from Canny")
+public class CannyBoardsResponseDTO {
+  @NotNull String message;
+  List<Board> boards;
```
Comment (sourabh-awashti) [inline platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyBoardsResponseDTO.java:27]: boards should also be notNull, as we can send empty List

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyClient.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+import okhttp3.MediaType;
+import okhttp3.MultipartBody;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import org.glassfish.jersey.media.multipart.FormDataContentDisposition;
+
+@Slf4j
+public class CannyClient {
```
Comment (sourabh-awashti) [inline platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyClient.java:28]: lets discuss on this offline

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyBoardsResponseDTO.java (hunk 6 above)
Comment (ashinsabu3) [reply platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyBoardsResponseDTO.java]: Added the annotation here

**Comment by ashinsabu3** [comment]
trigger uts

**Comment by ashinsabu3** [comment]
trigger ut4, ut9

**Comment by ashinsabu3** [comment]
trigger ut4

**Comment by ashinsabu3** [comment]
trigger ut4

**Comment by ashinsabu3** [comment]
trigger uts

**Comment by ashinsabu3** [comment]
trigger uts

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.inject.Inject;
+import java.util.*;
+import lombok.extern.slf4j.Slf4j;
+import okhttp3.Response;
+
+@Slf4j
+@OwnedBy(HarnessTeam.CDP)
+public class CannyHelper {
```
Comment (meenakshi-harness) [inline platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java:22]: Not sure if platform service is correct plan to have this @ankit-harness can you pls check once

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  @Inject CannyClient cannyClient;
+
+  public CannyBoardsResponseDTO getBoards() {
+    try (Response response = cannyClient.fetchBoardsResponseFromCanny()) {
+      if (!response.isSuccessful()) {
+        String bodyString = (null != response.body()) ? response.body().string() : "null";
+        log.error("Request to canny failed. Response body: {}", bodyString);
+        return CannyBoardsResponseDTO.builder().message("Request to canny failed.").build();
```
Comment (sourabh-awashti) [inline platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java:30]: we can send the message as failed to retrieve boards from canny

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyClient.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+      String jsonRequestBody = "{\"apiKey\":\"" + cannyConfig.token + "\"}";
+      RequestBody requestBody = RequestBody.create(MediaType.parse("application/json"), jsonRequestBody);
+      Request request =
+          new Request.Builder().url(cannyConfig.baseUrl + "/api/v1/boards/list").post(requestBody).build();
+
+      return okHttpClient.newCall(request).execute();
+
+    } catch (Exception e) {
+      log.error("Exception occurred while fetching boards at getBoards(): {}", e.getMessage(), e);
+      return new Response.Builder().code(500).message("Exception occurred while fetching boards from canny").build();
```
Comment (sourabh-awashti) [inline platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyClient.java:45]: can we verify if this exception is thrown then it is  handled properly in helper

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+package io.harness.resourcegroup.v1.remote.dto;
+
+import io.harness.annotations.dev.HarnessTeam;
+import io.harness.annotations.dev.OwnedBy;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.inject.Inject;
+import java.util.*;
```
Comment (sourabh-awashti) [inline platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java:16]: we should not use * imports

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        String bodyString = (null != response.body()) ? response.body().string() : "null";
+        log.error("Request to canny failed. Response body: {}", bodyString);
+        return CannyBoardsResponseDTO.builder().message("Request to canny failed.").build();
+      }
+
+      ObjectMapper objectMapper = new ObjectMapper();
+      // to avoid NullPointerException due to null response body
+      assert (response.body() != null);
+      JsonNode jsonResponse = objectMapper.readTree(response.body().byteStream());
+      JsonNode boardsNode = jsonResponse.get("boards");
```
Comment (sourabh-awashti) [inline platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java:37]: can we use string constants here in place of "boards" same with "id" and "name" below

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java (hunk 8 above)
Comment (ashinsabu3) [reply platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java:22]: The purpose of this api is to reroute FR tickets away from zendesk, the zendesk post creation API is there at https://github.com/harness/harness-core/blob/2670b6ff3bf6f95b6a9ee55a06c1ad088e58efec/platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/resource/ZendeskResource.java#L53-L59

@sourabh-awashti let me know if this api would have to be moved to other service

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java (hunk 9 above)
Comment (ashinsabu3) [reply platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java]: Addressed

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java (hunk 11 above)
Comment (ashinsabu3) [reply platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java]: Addressed

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java (hunk 12 above)
Comment (ashinsabu3) [reply platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java]: Addressed

File: platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java (hunk 8 above)
Comment (ankit-harness) [reply platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/CannyHelper.java:24]: @sourabh-awashti 
Platform Service is not the right place to have this as of today. Platform Service only has support for resource groups, notifications and audit and in future other core Platform functionalities. This change is not related to any of them. We can have it in NG manager for now.
Also, Zendesk endpoint is also added in the wrong place. It doesn't belong in Platform service alongside Resource Group endpoints. We should move that too to NG manager for now.

**Comment by ashinsabu3** [comment]
Closing this PR as endpoint will be moved to ng-manager

**Pull request closed without merging**